python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -147,11 +155,12 @@ def __init_subclass__(cls):
             assert (
                 len(invalid_operations) == 0
             ), f"Invalid requested operations: {invalid_operations}"
+
             for operation in valid_operations:
-                # Check if the operation is already defined so that subclasses
-                # can override the method if desired.
-                if not hasattr(cls, operation):
-                    cls._add_operation(operation)
+                if operation not in dir(cls):
```

Comment:
Why is `dir` preferred to `hasattr` here?

Reply:
A call to `hasattr()` here would trigger `getattr()` which in turn ends up calling the `__get__`  method for all subclasses `cls` at import time, something we definitely do not want.